File: golang/1054_rearrange_barcodes/rearrange_barcodes.go
/*
 * @Date: 2023-05-14
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-05-14
 * @FilePath: /algorithm/golang/1054_rearrange_barcodes/rearrange_barcodes.go
 */

// Package main ...
package main

import (
	"fmt"
	"reflect"
)

func rearrangeBarcodes(barcodes []int) []int {
	if len(barcodes) < 2 {
		return barcodes
	}

	counts := make(map[int]int)
	maxCount := 0
	for _, b := range barcodes {
		counts[b] = counts[b] + 1
		if counts[b] > maxCount {
			maxCount = counts[b]
		}
	}

	evenIndex := 0
	oddIndex := 1
	halfLength := len(barcodes) / 2
	res := make([]int, len(barcodes))
	for x, count := range counts {
		for count > 0 && count <= halfLength && oddIndex < len(barcodes) {
			res[oddIndex] = x
			count--
			oddIndex += 2
		}
		for count > 0 {
			res[evenIndex] = x
			count--
			evenIndex += 2
		}
	}
	return res
}

func main() {
	assert := func(a, b []int) {
		if !reflect.DeepEqual(a, b) {
			panic(fmt.Sprintf("%+v not equal %+v", a, b))
		}
	}

	{
		barcodes := []int{1, 1, 1, 2, 2, 2}
		ans := []int{2, 1, 2, 1, 2, 1}
		assert(rearrangeBarcodes(barcodes), ans)
	}

	{
		barcodes := []int{1, 1, 1, 1, 2, 2, 3, 3}
		ans := []int{2, 1, 2, 1, 3, 1, 3, 1}
		assert(rearrangeBarcodes(barcodes), ans)
	}
}

File: cpp/1054_rearrange_barcodes/rearrange_barcodes.cpp
/*
 * @Date: 2023-05-14
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-05-14
 * @FilePath: /algorithm/cpp/1054_rearrange_barcodes/rearrange_barcodes.cpp
 */

#include <cassert>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> rearrangeBarcodes(vector<int>& barcodes) {
    int length = barcodes.size();
    if (length < 2) {
      return barcodes;
    }

    unordered_map<int, int> counts;
    int maxCount = 0;
    for (int b : barcodes) {
      maxCount = max(maxCount, ++counts[b]);
    }

    int evenIndex = 0, oddIndex = 1, halfLength = length / 2;
    vector<int> res(length);
    for (auto& [x, cx] : counts) {
      while (cx > 0 && cx <= halfLength && oddIndex < length) {
        res[oddIndex] = x;
        cx--;
        oddIndex += 2;
      }
      while (cx > 0) {
        res[evenIndex] = x;
        cx--;
        evenIndex += 2;
      }
    }
    return res;
  }
};

int main() {
  {
    vector<int> barcodes{1, 1, 1, 2, 2, 2};
    vector<int> ans{2, 1, 2, 1, 2, 1};
    assert(Solution().rearrangeBarcodes(barcodes) == ans);
  }

  {
    vector<int> barcodes{1, 1, 1, 1, 2, 2, 3, 3};
    vector<int> ans{2, 1, 2, 1, 3, 1, 3, 1};
    assert(Solution().rearrangeBarcodes(barcodes) == ans);
  }
}